operator/identity/crd_gc.go

```diff
@@ -0,0 +1,184 @@
+	if err != nil {
+		igc.logger.WithError(err).Error("unable to list Cilium identities")
+	}
+	totalEntries := len(identities.Items)
+	deletedEntries := 0
+
+	timeNow := time.Now()
+	for _, identity := range identities.Items {
+		// The identity is definitely alive if there's a CE using it.
+		if watchers.HasCEWithIdentity(identity.Name) {
+			igc.heartbeatStore.MarkAlive(identity.Name, timeNow)
+			continue
+		}
+
+		if !igc.heartbeatStore.IsAlive(identity.Name) {
+			ts, ok := identity.Annotations[identitybackend.HeartBeatAnnotation]
+			if !ok {
+				log.WithFields(logrus.Fields{
+					logfields.Identity: identity.Name,
+					logfields.K8sUID:   identity.UID,
+				}).Info("Marking identity for later deletion")
+
+				if identity.Annotations == nil {
+					identity.Annotations = make(map[string]string)
+				}
+
+				identity.Annotations[identitybackend.HeartBeatAnnotation] = timeNow.Format(time.RFC3339Nano)
+				if err := igc.updateIdentity(ctx, &identity); err != nil {
+					log.WithError(err).
+						WithField(logfields.Identity, identity).
+						Error("Marking identity for later deletion")
+				}
+
+				continue
+			}
+
+			log.WithFields(logrus.Fields{
+				logfields.Identity: identity,
+			}).Debugf("Deleting unused identity; marked for deletion at %s", ts)
+
+			if err := igc.deleteIdentity(ctx, &identity); err != nil {
+				log.WithError(err).WithFields(logrus.Fields{
+					logfields.Identity: identity,
+				}).Error("Deleting unused identity")
+			} else {
+				deletedEntries++
+			}
+		}
+
+		// If Context was canceled we should break
+		if ctx.Err() != nil {
+			break
+		}
+	}
+
+	if igc.enableMetrics {
+		if ctx.Err() == nil {
+			igc.successfulRuns++
+			metrics.IdentityGCRuns.WithLabelValues(metrics.LabelValueOutcomeSuccess).Set(igc.successfulRuns)
+		} else {
+			igc.failedRuns++
+			metrics.IdentityGCRuns.WithLabelValues(metrics.LabelValueOutcomeFail).Set(igc.failedRuns)
+		}
+		aliveEntries := totalEntries - deletedEntries
+		metrics.IdentityGCSize.WithLabelValues(metrics.LabelValueOutcomeAlive).Set(float64(aliveEntries))
+		metrics.IdentityGCSize.WithLabelValues(metrics.LabelValueOutcomeDeleted).Set(float64(deletedEntries))
+	}
+
+	igc.heartbeatStore.GC()
+
+	return nil
+}
+
+// deleteIdentity deletes an identity. It includes the resource version and
+// will error if the object has since been changed.
+func (igc *GC) deleteIdentity(ctx context.Context, identity *v2.CiliumIdentity) error {
+	// Wait until we can delete an identity
+	if err := igc.rateLimiter.Wait(ctx); err != nil {
+		return err
+	}
+
+	if err := igc.clientset.Delete(
+		ctx,
+		identity.Name,
+		metav1.DeleteOptions{
+			Preconditions: &metav1.Preconditions{
+				UID:             &identity.UID,
+				ResourceVersion: &identity.ResourceVersion,
+			},
+		},
+	); err != nil {
+		log.WithError(err).Error("Unable to delete identity")
+		return err
+	}
+
+	log.WithField(logfields.Identity, identity.GetName()).Info("Garbage collected identity")
```

Comment:
Do we need to log this? Seems potentially quite spammy.

Reply:
It surely could be. Probably the right trade-off is to lower it to `Debug` level.